### ISSUE TITLE
Improve Follower-Leader transition performance

### ISSUE THREAD

**Issue by Zelldon** [camunda-cloud/zeebe#7515]
**Description**

This is part of the [ZEP 0007](https://github.com/zeebe-io/enhancements/blob/master/ZEP007-build-state-on-followers.md)

Main goal of this ZEP is to reduce the time the new Leader needs to start processing. We want to reduce the impact on process instance execution latency, which a fail-over has.

This can be shown in the with the following Seq Diagram

![](https://github.com/zeebe-io/enhancements/blob/master/ZEP007/leaderTransition.png?raw=true)

The ZEP targets the range the ZeebePartition is called with the Leader Role event until the StreamProcessor starts processing. 

Part of this issue is to stop recreating the runtime folder from the last snapshot, ideally we also stop to recreate the Database object and reuse it from the Follower. **Important:** We need to make sure that we still reset the runtime db on transition from Leader to follower.

Blocked by https://github.com/camunda-cloud/zeebe/issues/7513

**Comment by deepthidevaki** [comment]
PR https://github.com/camunda-cloud/zeebe/pull/7739 introduced the new Transition logic in which we can transition to a new role without re-installing all services. 

The plan is to move each PartitionStep to the new interface PartitionTransitionStep. While moving we can conditionally re-installs the services. For ZeebeDb, we should then re-install only when transitioning from leader to follower. 

The steps to achieve this

- [ ] Implement all PartitionSteps using the new interface PartitionTransitionStep
     - [ ] StreamProcessorStep
     - [ ] ZeebeDbPartitionStep
     - [ ] ExporterDirectorPartitionStep
     - [ ] LogStoragePartitionStep
     - [ ] LogStreamPartitionStep
     - [ ] SnapshotDirectorPartitionStep
 - [ ] Switch to the new transition logic
     - [ ] Use the feature flag to enable the new transition logic and test it
     - [ ] Remove all implementations of the old transition logic

**Comment by Zelldon** [comment]
Please close this after we have the benchmark results @deepthidevaki

**Comment by deepthidevaki** [comment]
Oops. Sorry. This issue is not supposed to be closed.

**Comment by deepthidevaki** [comment]
I'm currently testing the new transition (branch `dd-use-new-transition`). 

I started two cluster - one with old transition, the other with new new transition. And periodically disconnected network to induce leader changes.

The following figure shows the reprocessing time. With the new transition it is only a few milliseconds as it does not have to reprocess from the last snapshot. The old transition can take a few seconds. In some benchmarks I have observed it took upto a few minuted depending on how old is the latest snapshot.
![image](https://user-images.githubusercontent.com/1997478/133763505-399a21f3-fca9-46e6-8c4d-be201541321a.png)


The following shows Leader transition duration ( = time needed to install leader services).

![image](https://user-images.githubusercontent.com/1997478/133764959-0bb664ba-be84-465d-8f61-5e3e385901a3.png)
For old transition it is most of the time < 200ms, but once it reached 3s when the snapshot was bigger.

![image](https://user-images.githubusercontent.com/1997478/133765466-9a9859ed-b686-4b83-ad71-73d754236783.png)
For new transition it is < 100ms. The initial large value (~1s) is the initial transition immediately after the startup.
In total with the new transition from follower to leader requires ~100ms, while previously it was a few seconds upto a few minutes.

Note that this time does not include the election. This only shows how much time is needed for a leader to start accepting requests after it is elected.
